src/hardware/uni-t-ut8802e/protocol.c

```diff
@@ -0,0 +1,430 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include "math.h"
+
+#include "protocol.h"
+
+/*
+ * Developer notes on the UT8802E protocol:
+ * - Serial communication over HID CP2110 converter USB-UART.
+ * - UART frame format 8n1 at 9600 bps.
+ * - DMM packet starts with a magic marker, functionality, value, etc.
+ * - UT8802E only sends a packet with measurement. Example of frame:
+ *   Frame | func | value    | comma | settings | checksum |
+ *   ac    | 1b   | 45 01 00 | 33	 | 04  	    | 44       |
+ * - The frame to send a command to the multimeter is unknown.
+*/
+
+static const int8_t range_volt[]  = {-3,  0};
+static const int8_t range_amp[]   = {-6, -3,  0};
+static const int8_t range_ohm[]   = {0,   3,  6};
+static const int8_t range_farad[] = {-9, -6, -3};
+static const int8_t range_hz[]    = {0,   3,  6};
+
+static int ut8802e_feedbuff_initialization(struct feed_buffer *buff) {
+    memset(buff, 0, sizeof(*buff));
+    memset(&buff->packet, 0, sizeof(buff->packet));
+    sr_analog_init(&buff->analog, &buff->encoding,
+                   &buff->meaning, &buff->spec, 0);
+    buff->analog.meaning->mq = 0;
+    buff->analog.meaning->mqflags = 0;
+    buff->analog.meaning->unit = 0;
+    buff->analog.meaning->channels = 0;
+    buff->analog.encoding->unitsize = sizeof(buff->main_value);
+    buff->analog.encoding->digits = 4;
+    buff->analog.spec->spec_digits = 4;
+    buff->analog.num_samples = 1;
+    buff->analog.data = &buff->main_value;
+    buff->packet.type = SR_DF_ANALOG;
+    buff->packet.payload = &buff->analog;
+    buff->analog.encoding->is_float = TRUE;
+    return SR_OK;
+}
+
+static int process_packet(struct sr_dev_inst *sdi, uint8_t *pkt, size_t len) {
+    struct dev_context *devc;
+    struct ut8802e_info *info;
+    const uint8_t *payload;
+    size_t checksum_dlen;
+
+    uint8_t v8, got_magic_frame, got_checksum, want_checksum;
+    uint16_t clc_checksum;
+    uint32_t v32;
+    uint8_t got_length, settings;
+    float new_value;
+
+    struct feed_buffer feedbuff;
+
+    int ret;
+
+    devc = sdi ? sdi->priv : NULL;
+    info = devc ? &devc->info : NULL;
+
+    if (len < 1 * sizeof(uint8_t)) {
+        sr_spew("Wrong packet");
+        return SR_ERR_DATA;
+    }
+
+    got_magic_frame = R8(&pkt[0]);
+
+    if (got_magic_frame != FRAME_MAGIC) {
+        sr_spew("Wrong frame packet");
+        return SR_ERR_DATA;
+    }
+
+    got_length = 7;
+    if (got_length != len - sizeof(uint8_t)) {
+        sr_spew("Wrong lenght of packet");
+        return SR_ERR_DATA;
+    }
+
+    payload = &pkt[sizeof(uint8_t)];
+
+    clc_checksum = 0;
+    checksum_dlen = got_length;
+    for (uint32_t i = 0; i < checksum_dlen; i++) {
+        clc_checksum = clc_checksum + pkt[i];
+    }
+    want_checksum = clc_checksum;
+
+    if (clc_checksum > 256 && want_checksum >= 128)
+        want_checksum = want_checksum - 128;
+
```

Comment:
The checksum is calculated over the first 8 bytes including the FRAME_MAGIC byte (0xAC).  If the checksum is in the range 0xAC - 0xFF, i.e. the sum of the remaining bytes is less than 0x54, the checksum should be  found in pkt[8] with bit 7 set. But if the checksum is 0x100 or more, bit 7 of the checksum will be reset.
This is a strange definition.

Reply:
Thank you very much. I made a mistake here. I fixed it now. Bit 7 is always cleared.
0xac 0x04 0x10 0x20 0x01 0x33 0x04 0x18 -> Sum of bytes 0x118,
0xac 0x06 0x00 0x00 0x00 0x31 0x04 0x67 (103) -> Sum of bytes 0xE7 (231) 
0xac 0x04 0x88 0x99 0x01 0x33 0x04 0x09 -> Sum of bytes 0x209
0xac 0x04 0x89 0x69 0x01 0x33 0x04 0x5a (90)-> Sum of bytes 0x1DA (474)